backends/graphics/openglsdl/openglsdl-graphics.cpp

```diff
@@ -352,22 +348,27 @@ bool OpenGLSdlGraphicsManager::loadVideoMode(uint requestedWidth, uint requested
 	// Fetch current desktop resolution and determining max. width and height
 	Common::Rect desktopRes = _window->getDesktopResolution();
 
-	if (_windowIsMaximized == true) {
+	// Determine current desktop aspect ratio
+	float ratio = (float)desktopRes.width() / (float)desktopRes.height();
+
+#if SDL_VERSION_ATLEAST(2, 0, 0)
+	bool _isMaximized = (SDL_GetWindowFlags(_window->getSDLWindow()) & SDL_WINDOW_MAXIMIZED);
+	if (_isMaximized && ConfMan.hasKey("window_maximized_width", Common::ConfigManager::kApplicationDomain) && ConfMan.hasKey("window_maximized_height", Common::ConfigManager::kApplicationDomain)) {
 		// Set the window size to the values stored when the window was maximized
-		// for the last time. We also need to reset any scaling here.
+		// for the last time.
 		requestedWidth  = ConfMan.getInt("window_maximized_width", Common::ConfigManager::kApplicationDomain);
 		requestedHeight = ConfMan.getInt("window_maximized_height", Common::ConfigManager::kApplicationDomain);
 
-	} else if (ConfMan.hasKey("last_window_width", Common::ConfigManager::kApplicationDomain) && ConfMan.hasKey("last_window_height", Common::ConfigManager::kApplicationDomain)) {
+	} else if (!_isMaximized && ConfMan.hasKey("last_window_width", Common::ConfigManager::kApplicationDomain) && ConfMan.hasKey("last_window_height", Common::ConfigManager::kApplicationDomain)) {
 		// Restore previously stored window dimensions.
 		requestedWidth  = ConfMan.getInt("last_window_width", Common::ConfigManager::kApplicationDomain);
 		requestedHeight = ConfMan.getInt("last_window_height", Common::ConfigManager::kApplicationDomain);
 
 	} else {
 		// Set the basic window size based on the desktop resolution
 		// since we have no values stored, e.g. on first launch.
-		requestedWidth  = desktopRes.width()  * 0.3f;
-		requestedHeight = desktopRes.height() * 0.4f;
+		requestedWidth  = desktopRes.width()  * (1.0f / 3.0f) * 1.5f / ratio;
+		requestedHeight = desktopRes.height() * (1.0f / 4.0f) * 1.5f;
 
```

Comment:
Isn't  that more complicated than needed?
Unless I am mistaken, your code is equivalent to:
```c++
requestedHeight = desktopRes.height() * 3 / 8;
requestedWidth = requestedHeight * 4 / 3;
```
Note: I also changed it to use integer computations to avoid float to int conversions (since `requestedWidth` and `requestedHeight` are integers).

But you could try to handle a bit better small screens as well with a minimum width of 640 (or the screen width if smaller):
```c++
requestedWidth  = MAX(desktopRes.width() / 2, MIN(640, desktopRes.width()));
requestedHeight = MIN(requestedWidth * 3 / 4, desktopRes.height());
```

I am however wondering if we should really ignore the `requestedWidth` and `requestedHeight` passed to the function. These might for example be the size of the game when starting a game (and if we start the game from the command line with an empty ini file, we will hit this case). So I am thinking that we should use a size that:
 - Has the same aspect ratio than the provided ones (rather than force 4:3).
 - Is at least as big as the given size (provided it fits on the screen).

This last lines also applies to the case where we have a `last_window_width` and `last_window_height` in ConfMan. I think they should only be used if they are bigger than the requested size.

One suggestion for the default (if the window is not maximized and there is no `last_window_width` and `last_window_height` in ConfMan) would be to use the multiple of the requested size passed to the function that is closest to the half screen size, with a minimum of 640:
```c++
frac_t ratio = intToFrac(requestedWidth) / requestedHeight;
int scale = MIN(1, (desktopRes.width() / 2 + requestedWidth / 2) / requestedWidth);
requestedWidth = MAX(requestedWidth * scale, MIN(640, desktopRes.width()));
requestedHeight = MIN(intToFrac(requestedWidth) / ratio, desktopRes.height());

Reply:
> Isn't that more complicated than needed?
> Unless I am mistaken, your code is equivalent to:
> 
> ```c++
> requestedHeight = desktopRes.height() * 3 / 8;
> requestedWidth = requestedHeight * 4 / 3;
> ```
> 
> Note: I also changed it to use integer computations to avoid float to int conversions (since `requestedWidth` and `requestedHeight` are integers).
> 
> But you could try to handle a bit better small screens as well with a minimum width of 640 (or the screen width if smaller):
> 
> ```c++
> requestedWidth  = MAX(desktopRes.width() / 2, MIN(640, desktopRes.width()));
> requestedHeight = MIN(requestedWidth * 3 / 4, desktopRes.height());
> ```
> 

Yes, indeed this is a great idea - thank you! I'll adapt the code accordingly.

> I am however wondering if we should really ignore the `requestedWidth` and `requestedHeight` passed to the function. These might for example be the size of the game when starting a game (and if we start the game from the command line with an empty ini file, we will hit this case). So I am thinking that we should use a size that:
> 
>     * Has the same aspect ratio than the provided ones (rather than force 4:3).
> 
>     * Is at least as big as the given size (provided it fits on the screen).
> 
> 
> This last lines also applies to the case where we have a `last_window_width` and `last_window_height` in ConfMan. I think they should only be used if they are bigger than the requested size.

I'm not sure about this since the main idea I had while implementing this originally was to avoid any "size switching" and only stick to the dimensions either calculated initially or set by the user. I see that we shouldn't disrespect the `requestedWidth` and `requestedHeight` values, but for now, the only downside (unless you use the "Stretch to window" option) is that you might notice some black borders around games that have an "odd" resolution.

> 
> One suggestion for the default (if the window is not maximized and there is no `last_window_width` and `last_window_height` in ConfMan) would be to use the multiple of the requested size passed to the function that is closest to the half screen size, with a minimum of 640:
> 
> ```c++
> frac_t ratio = intToFrac(requestedWidth) / requestedHeight;
> int scale = MIN(1, (desktopRes.width() / 2 + requestedWidth / 2) / requestedWidth);
> requestedWidth = MAX(requestedWidth * scale, MIN(640, desktopRes.width()));
> requestedHeight = MIN(intToFrac(requestedWidth) / ratio, desktopRes.height());
> ```

The main issue here is that `desktopRes.*` returns the full desktop resolution, ignoring any space used by e.g. the Windows taskbar or window borders. When using this code, this leads to the image being drawn too large on high-res games unfortunately, and especially on lower resolutions.

Looking at the upcoming release and feature-freeze, I'm worried to run out of time here. The current upstream `master` branch is clearly broken here, so for me it has highest priority to improve the situation, even if the solution is not perfect :(

Reply:
> > I am however wondering if we should really ignore the `requestedWidth` and `requestedHeight` passed to the function. These might for example be the size of the game when starting a game (and if we start the game from the command line with an empty ini file, we will hit this case). So I am thinking that we should use a size that:
> > 
> > * Has the same aspect ratio than the provided ones (rather than force 4:3).
> > * Is at least as big as the given size (provided it fits on the screen).
> >   
> > This last lines also applies to the case where we have a `last_window_width` and `last_window_height` in ConfMan. I think they should only be used if they are bigger than the requested size.
> 
> I'm not sure about this since the main idea I had while implementing this originally was to avoid any "size switching" and only stick to the dimensions either calculated initially or set by the user. I see that we shouldn't disrespect the `requestedWidth` and `requestedHeight` values, but for now, the only downside (unless you use the "Stretch to window" option) is that you might notice some black borders around games that have an "odd" resolution.

I agree that at this point we can aim for something that at least works and look a improving it further later. The only point that I am worried about is my second point. If your current ScummVM window size is 640x480 and you are starting a game with a 800x600 resolution for example, I think we want to resize the window to be at least 800x600. But maybe I am missing something and this is already happening somewhere else?

Reply:
Ah, no, this is not happening. If you start with a 640x480 window, then your game will be downscaled.